cmake/CeleritasLibrary.cmake

```diff
@@ -173,111 +176,105 @@ function(celeritas_add_library target)
     CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
   )
+  set_target_properties(${target} PROPERTIES
+    ${_all_props}
+    CELERITAS_CUDA_LIBRARY_TYPE Shared
+    CUDA_RESOLVE_DEVICE_SYMBOLS OFF # We really don't want nvlink called.
+  )
+
+  ## STATIC ##
 
-  if(NOT _static_build)
+  if(_staticsuf)
+    add_library(${target}${_staticsuf} STATIC
+      $<TARGET_OBJECTS:${target}_objects>
+    )
     set_target_properties(${target}${_staticsuf} PROPERTIES
-      LINKER_LANGUAGE CUDA
-      CUDA_SEPARABLE_COMPILATION ON
-      CUDA_RUNTIME_LIBRARY ${_cudaruntime_requested_type}
-      # CUDA_RESOLVE_DEVICE_SYMBOLS OFF # Default for static library
+      ${_all_props}
       CELERITAS_CUDA_LIBRARY_TYPE Static
-      CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
-      CELERITAS_CUDA_MIDDLE_LIBRARY ${target}${_midsuf}
-      CELERITAS_CUDA_STATIC_LIBRARY ${target}${_staticsuf}
-      CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
     )
   endif()
 
+  ## FINAL (dlink) ##
+
+  # We need to use a dummy file as a library (per cmake) needs to contains
+  # at least one source file.  The real content of the library will be
+  # the cmake_device_link.o resulting from the execution of `nvcc -dlink`
+  # Also non-cuda related test, for example `gtest_detail_Macros`,
+  # will need to be linked again libceleritas_final while a library
+  # that the depends on and that uses Celeritas::Core (for example
+  # libCeleritasTest.so) will need to be linked against `libceleritas`.
+  # If both the middle and `_final` contains the `.o` files we would
+  # then have duplicated symbols .  If both the middle and `_final`
+  # library contained the result of `nvcc -dlink` then we would get
+  # conflicting but duplicated *weak* symbols and here the symptoms
+  # will be a crash during the cuda library initialization or a failure to
+  # launch some kernels rather than a link error.
+  celeritas_generate_empty_cu_file(_emptyfilename ${target})
+  add_library(${target}_final ${_lib_requested_type} ${_emptyfilename})
   set_target_properties(${target}_final PROPERTIES
-    LINKER_LANGUAGE CUDA
-    CUDA_RESOLVE_DEVICE_SYMBOLS ON
-    CUDA_SEPARABLE_COMPILATION ON
-    CUDA_RUNTIME_LIBRARY ${_cudaruntime_requested_type}
-    # CUDA_RESOLVE_DEVICE_SYMBOLS ON # Default for shared library
+    ${_all_props}
     CELERITAS_CUDA_LIBRARY_TYPE Final
-    CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
-    CELERITAS_CUDA_STATIC_LIBRARY ${target}${_staticsuf}
-    CELERITAS_CUDA_MIDDLE_LIBRARY ${target}${_midsuf}
-    CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
-  )
-
-  target_link_libraries(${target}_final
-    PUBLIC ${target}${_midsuf}
+    CUDA_RESOLVE_DEVICE_SYMBOLS ON
   )
-
+  target_link_libraries(${target}_final PUBLIC ${target})
   target_link_options(${target}_final
-    PRIVATE
-    $<DEVICE_LINK:$<TARGET_FILE:${target}${_staticsuf}>>
+    PRIVATE $<DEVICE_LINK:$<TARGET_FILE:${target}${_staticsuf}>>
   )
-
-  add_dependencies(${target}_final ${target}${_midsuf})
```

Comment:
Was this dependency removed intentionally?  There is seemingly not explanation in the commit ("slight refactor of CeleritasLibrary").  I plan on re-adding it as part of the sync with the unofficial cousin/upstream (https://github.com/drbenmorgan/CudaRDCExercise)

Reply:
Answering my own question:  It was intentionally removed as it is redundant with the `target_link_libraries` on line 273

Reply:
Yep, exactly.